File: src/fpga/software/host/check_output.py
#!/usr/bin/env python3
"""Compare a baremetal VTA output dump against the functional simulator's
compiler_output/final_output.bin (the source of truth).

The board writes the output block-tiled; this detiles it to NCHW and diffs it
against final_output.bin (already NCHW).

The block layout mirrors output_tensor()/unsplit() in
src/simulators/functional_simulator/include/cpu_functions.h: blocks are laid out
row-major over (N/B row-blocks, C/B col-blocks); within a block element [rr][cc]
maps to (spatial row rr, channel cc). The matrix is [N=H*W rows, C cols].

Usage:
    python3 host/check_output.py <baremetal_out.bin> \\
        [--ref <compiler_output>/final_output.bin] \\
        [--comp-dir <compiler_output>] [--config-json <cfg>.json] \\
        [--shape C,H,W] [--block-size N]
"""

import argparse
import os
import sys

import numpy as np

import gen_nn_baremetal as gen

_HERE = os.path.dirname(__file__)
_DEF_COMP = os.path.join(_HERE, "..", "..", "..", "..", "compiler_output")
_DEF_CFG = os.path.join(_HERE, "..", "..", "..", "..", "config", "vta_config.json")


def _ceil_div(a: int, b: int) -> int:
    return (a + b - 1) // b


def detile(raw: np.ndarray, C: int, H: int, W: int, B: int) -> np.ndarray:
    """Block-tiled [N/B][C/B][B][B] -> NCHW flat [C][H*W]. Drops padding cells."""
    N = H * W
    Cb = _ceil_div(C, B)
    expect = _ceil_div(N, B) * Cb * B * B
    if raw.size < expect:
        raise ValueError(f"raw output too small: {raw.size} < expected tiled {expect}")
    out = np.zeros(C * N, dtype=raw.dtype)
    for n in range(N):
        rb, rr = divmod(n, B)
        for c in range(C):
            cb, cc = divmod(c, B)
            out[c * N + n] = raw[(rb * Cb + cb) * B * B + rr * B + cc]
    return out


def report_diff(name: str, a: np.ndarray, b: np.ndarray) -> bool:
    """Diff two arrays; print stats; return True if identical."""
    n = min(a.size, b.size)
    if a.size != b.size:
        print(f"[{name}] SIZE DIFF: got {a.size} vs ref {b.size} (comparing first {n})")
    a32 = a[:n].astype(np.int32)
    b32 = b[:n].astype(np.int32)
    diff = a32 - b32
    nz = np.flatnonzero(diff)
    if a.size == b.size and nz.size == 0:
        print(f"[{name}] MATCH ({n} elements identical)")
        return True
    first = int(nz[0]) if nz.size else -1
    print(f"[{name}] MISMATCH: {nz.size}/{n} differ "
          f"({100.0 * nz.size / n:.3f}%), max|diff|={int(np.abs(diff).max())}, "
          f"first @ idx {first}")
    if first >= 0:
        lo, hi = max(0, first - 4), min(n, first + 12)
        print(f"        idx[{lo}:{hi}] got = {a32[lo:hi].tolist()}")
        print(f"        idx[{lo}:{hi}] ref = {b32[lo:hi].tolist()}")
    return False


def main() -> int:
    p = argparse.ArgumentParser(description="Compare baremetal VTA output to final_output.bin")
    p.add_argument("baremetal_out", help="raw output dump from the board (int8, block-tiled)")
    p.add_argument("--ref", default=None, help="reference NCHW bin (default <comp-dir>/final_output.bin)")
    p.add_argument("--comp-dir", default=_DEF_COMP)
    p.add_argument("--config-json", default=_DEF_CFG)
    p.add_argument("--shape", default=None, help="C,H,W override (else read from dependency.csv)")
    p.add_argument("--block-size", type=int, default=None)
    args = p.parse_args()

    cfg = gen.load_config_params(args.config_json, args.block_size)
    B = cfg.block_size

    if args.shape:
        C, H, W = (int(x) for x in args.shape.split(","))
    else:
        dep = gen.load_dependency_csv(os.path.join(args.comp_dir, "dependency.csv"))
        ld = dep.layers.get(dep.output_layer)
        if ld is None:
            sys.exit(f"ERROR: output layer '{dep.output_layer}' not in dependency.csv; pass --shape")
        C, H, W = ld.out_ch, ld.out_h, ld.out_w

    ref_path = args.ref or os.path.join(args.comp_dir, "final_output.bin")
    raw = np.fromfile(args.baremetal_out, dtype=np.int8)
    ref = np.fromfile(ref_path, dtype=np.int8)

    print(f"shape C,H,W = {C},{H},{W}  block={B}")
    print(f"baremetal raw: {raw.size} B   reference (NCHW): {ref.size} B   ({ref_path})\n")

    ok = report_diff("detiled vs NCHW ref", detile(raw, C, H, W, B), ref)
    print("\nRESULT:", "PASS - matches the functional simulator." if ok
          else "FAIL - output diverges from the functional simulator.")
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: vta/fpga/src/main/scala/host/checkers/CheckOutput.scala
package fpga.host.checkers

import vta.parsers.Dependency._
import scopt.OParser
import fpga.host.parsers.ConfigParser

/** Compare a baremetal VTA output dump against the functional simulator's
  * final_output.bin (the source of truth).
  *
  * The board writes the output block-tiled; [[detile]] converts it to NCHW and
  * [[reportDiff]] diffs it against the reference.
  *
  * The block layout mirrors output_tensor()/unsplit() in
  * src/simulators/functional_simulator/include/cpu_functions.h: blocks are laid
  * out row-major over (N/B row-blocks, C/B col-blocks); within a block element
  * [rr][cc] maps to (spatial row rr, channel cc). The matrix is [N=H*W rows, C
  * cols].
  *
  * Translated from host/check_output.py.
  */
object CheckOutput {

  private def ceilDiv(a: Int, b: Int): Int = (a + b - 1) / b

  /** Block-tiled [N/B][C/B][B][B] -> NCHW flat [C][H*W]. Drops padding cells.
    *
    * Mirrors the Python detile() function exactly: same index formula
    * `(rb*Cb+cb)*B*B + rr*B + cc`.
    */
  def detile(raw: Array[Byte], C: Int, H: Int, W: Int, B: Int): Array[Byte] = {
    val N = H * W
    val Cb = ceilDiv(C, B)
    val expect = ceilDiv(N, B) * Cb * B * B
    require(
      raw.length >= expect,
      s"raw output too small: ${raw.length} < expected tiled $expect"
    )
    val out = new Array[Byte](C * N)
    for (n <- 0 until N) {
      val rb = n / B
      val rr = n % B
      for (c <- 0 until C) {
        val cb = c / B
        val cc = c % B
        out(c * N + n) = raw((rb * Cb + cb) * B * B + rr * B + cc)
      }
    }
    out
  }

  /** Diff two byte arrays as signed ints; print stats; return true if
    * identical.
    *
    * Mirrors the Python report_diff() function exactly, including the SIZE DIFF
    * / MATCH / MISMATCH output format and the context window (first-4 ..
    * first+12).
    */
  def reportDiff(name: String, a: Array[Byte], b: Array[Byte]): Boolean = {
    val n = math.min(a.length, b.length)
    if (a.length != b.length)
      println(
        s"[$name] SIZE DIFF: got ${a.length} vs ref ${b.length} (comparing first $n)"
      )
    val a32 = a.take(n).map(_.toInt)
    val b32 = b.take(n).map(_.toInt)
    val diff = a32.zip(b32).map { case (x, y) => x - y }
    val nz = diff.zipWithIndex.collect { case (v, i) if v != 0 => i }

    if (a.length == b.length && nz.isEmpty) {
      println(s"[$name] MATCH ($n elements identical)")
      return true
    }

    val first = if (nz.nonEmpty) nz(0) else -1
    val pct = 100.0 * nz.length / n
    val maxDiff = if (diff.nonEmpty) diff.map(math.abs).max else 0
    println(
      f"[$name] MISMATCH: ${nz.length}/$n differ ($pct%.3f%%)," +
        s" max|diff|=$maxDiff," +
        s" first @ idx $first"
    )
    if (first >= 0) {
      val lo = math.max(0, first - 4)
      val hi = math.min(n, first + 12)
      println(s"        idx[$lo:$hi] got = ${a32.slice(lo, hi).toList}")
      println(s"        idx[$lo:$hi] ref = ${b32.slice(lo, hi).toList}")
    }
    false
  }

  private case class Opts(
    baremetalOut: String = "",
    compDir: String = "compiler_output",
    simDir: String = "simulators_output",
    configJson: Option[String] = None,
    shape: Option[String] = None,
    ref: Option[String] = None,
    detiled: Boolean = false
  )

  private val argParser: OParser[_, Opts] = {
    val b = OParser.builder[Opts]
    import b._
    OParser.sequence(
      programName("checkOutput"),
      head(
        "checkOutput",
        "compare a baremetal VTA output dump against the fsim reference"
      ),
      arg[String]("<baremetal_output>")
        .action((x, c) => c.copy(baremetalOut = x))
        .text("baremetal output file (block-tiled or pre-detiled)"),
      opt[String]("comp-dir")
        .action((x, c) => c.copy(compDir = x))
        .text("compiler output directory (default compiler_output)"),
      opt[String]("sim-dir")
        .action((x, c) => c.copy(simDir = x))
        .text("simulator output directory (default simulators_output)"),
      opt[String]("config-json")
        .action((x, c) => c.copy(configJson = Some(x)))
        .text("path to vta_config.json (default: Chisel DefaultPynqConfig)"),
      opt[String]("shape")
        .action((x, c) => c.copy(shape = Some(x)))
        .text("output shape C,H,W (inferred from dependency.csv if omitted)"),
      opt[String]("ref")
        .action((x, c) => c.copy(ref = Some(x)))
        .text("reference binary (default <sim-dir>/final_output.bin)"),
      opt[Unit]("detiled")
        .action((_, c) => c.copy(detiled = true))
        .text("treat input as already NCHW (skip detile step)")
    )
  }

  def main(args: Array[String]): Unit =
    OParser.parse(argParser, args, Opts()) match {
      case Some(o) =>
        val compDir = o.compDir
        val detiledFlag = o.detiled

        val cfg = ConfigParser.load(o.configJson.orNull)
        val blk = cfg.blockSize

        val (outC, outH, outW) = o.shape match {
          case Some(shapeStr) =>
            val parts = shapeStr.split(",")
            (parts(0).toInt, parts(1).toInt, parts(2).toInt)
          case None =>
            val dep = loadDependencyInfo(s"$compDir/dependency.csv")
            dep.layers.get(dep.outputLayer) match {
              case None =>
                System.err.println(
                  s"ERROR: output layer '${dep.outputLayer}' not in dependency.csv; pass --shape"
                )
                sys.exit(1)
              case Some(ld) => (ld.outCh, ld.outH, ld.outW)
            }
        }

        val refPath = o.ref.getOrElse(s"${o.simDir}/final_output.bin")
        val raw = os.read.bytes(os.Path(o.baremetalOut, os.pwd))
        val ref = os.read.bytes(os.Path(refPath, os.pwd))

        println(s"shape C,H,W = $outC,$outH,$outW  block=$blk")
        val layout = if (detiledFlag) "NCHW (pre-detiled)" else "block-tiled"
        println(
          s"baremetal raw: ${raw.length} B [$layout]   " +
            s"reference (NCHW): ${ref.length} B   ($refPath)\n"
        )

        val got = if (detiledFlag) raw else detile(raw, outC, outH, outW, blk)
        val label =
          if (detiledFlag) "NCHW vs NCHW ref" else "detiled vs NCHW ref"
        val ok = reportDiff(label, got, ref)
        val result =
          if (ok) "PASS - matches the functional simulator."
          else "FAIL - output diverges from the functional simulator."
        println("\nRESULT: " + result)
        sys.exit(if (ok) 0 else 1)
      case _ => sys.exit(1)
    }
}
